state/watcher.go

```diff
@@ -1096,150 +1095,291 @@ func (w *relationUnitsWatcher) loop() (err error) {
 		// If no members criteria is specified, use the filter
 		// to reject any unsuitable initial elements.
-		if w.members == nil && w.filter != nil && !w.filter(doc.Id) {
+		if w.members == nil && !w.filter(doc.Id) {
 			continue
 		}
 		id := w.backend.localID(doc.Id)
 		ids.Add(id)
-		if doc.Life != Dead {
-			w.knownLifeStatus[id] = lifeStatus{doc.Life, doc.Status}
-		}
+		w.knownStatus[id] = doc.Status
 	}
-	return ids, nil
+	return ids, iter.Close()
 }
 
-func (w *relationStatusWatcher) merge(keys set.Strings, updates map[interface{}]bool) error {
-	coll, closer := w.db.GetCollection(relationsC)
+func (w *statusWatcher) merge(ids set.Strings, updates map[interface{}]bool) (err error) {
+	statusColl, closer := w.db.GetCollection(statusesC)
 	defer closer()
 
-	// Separate keys into those thought to exist and those known to be removed.
+	// Separate ids into those existing and those known to be removed.
 	var changed []string
-	latestLifeStatus := make(map[string]lifeStatus)
+	gone := make(set.Strings)
+	latest := make(map[string]status.Status)
 	for docID, exists := range updates {
 		switch docID := docID.(type) {
 		case string:
 			if exists {
 				changed = append(changed, docID)
 			} else {
-				key := w.backend.localID(docID)
-				latestLifeStatus[key] = lifeStatus{Dead, status.Broken}
+				gone.Add(w.backend.localID(docID))
 			}
 		default:
 			return errors.Errorf("id is not of type string, got %T", docID)
 		}
 	}
 
-	// Collect life and status from keys thought to exist. Any that don't actually
-	// exist are ignored (we'll hear about them in the next set of updates --
-	// all that's actually happened in that situation is that the watcher
-	// events have lagged a little behind reality).
-	iter := coll.Find(bson.D{{"_id", bson.D{{"$in", changed}}}}).Select(lifeStatusFields).Iter()
-	var doc lifeStatusDoc
+	iter := statusColl.Find(bson.D{{"_id", bson.D{{"$in", changed}}}}).Select(statusFields).Iter()
+	var doc statusValueDoc
 	for iter.Next(&doc) {
-		key := w.backend.localID(doc.Id)
-		latestLifeStatus[key] = lifeStatus{doc.Life, doc.Status}
+		latest[w.backend.localID(doc.Id)] = doc.Status
 	}
 	if err := iter.Close(); err != nil {
 		return err
 	}
 
-	// Now correlate the changes.
-	for key, newLifeStatus := range latestLifeStatus {
-		gone := newLifeStatus.life == Dead
-		oldLifeStatus, known := w.knownLifeStatus[key]
+	// Add to ids any whose status value is known to have changed.
+	// Don't send events for deleted/removed status entities.
+	for id, newStatus := range latest {
+		gone := gone.Contains(id)
+		oldStatus, known := w.knownStatus[id]
 		switch {
-		case known && gone:
-			delete(w.knownLifeStatus, key)
 		case !known && !gone:
-			w.knownLifeStatus[key] = newLifeStatus
-		case known && (newLifeStatus.life != oldLifeStatus.life || newLifeStatus.status != oldLifeStatus.status):
-			w.knownLifeStatus[key] = newLifeStatus
+			w.knownStatus[id] = newStatus
+		case known && newStatus != oldStatus:
```

Comment:
this case shouldn't fire if "gone" is true, right? I think you want to swap this with the one below

Reply:
sounds right